netket/graph/custom_graph.py

```diff
@@ -0,0 +1,99 @@
+from abstract_graph import PyAbstractGraph
+
+import numpy as _np
+import networkx as _nx
+
+class PyCustomGraph(PyAbstractGraph):
+	r"""A custom graph, specified by a list of edges and optionally colors."""
+
+	def __init__(self, edges=[]):
+		"""
+		Constructs a new graph given a list of edges.
+
+		Args:
+		    edges: If `edges` has elements of type `Tuple[int, int]` it is treated
+		        as a list of edges. Then each element `(i, j)` means a connection
+		        between sites `i` and `j`. It is assumed that `0 <= i <= j`. Also,
+		        `edges` should contain no duplicates. If `edges` has elements of
+		        type `Tuple[int, int, int]` each element `(i, j, c)` represents an
+		        edge between sites `i` and `j` colored into `c`. It is again assumed
+		        that `0 <= i <= j` and that there are no duplicate elements in `edges`.
+
+
+		Examples:
+		    A 10-site one-dimensional lattice with periodic boundary conditions can be
+		    constructed specifying the edges as follows:
+
+		    >>> import netket
+		    >>> g=netket.graph.PyCustomGraph([[i, (i + 1) % 10] for i in range(10)])
+		    >>> print(g.n_sites)
+		    10
+		"""
+
+		assert isinstance(edges, list)
+		type_condition = [isinstance(edge, list) or isinstance(edge, tuple) for edge in edges]
+		if False in type_condition:
+			raise ValueError("edges must be a list of lists or tuples")
+
+		edges_array = _np.array(edges, dtype=_np.int32)
+		if edges_array.ndim != 2:
+			raise ValueError("edges must be a list of lists or tuples of the same length (2 or 3)")
+
+		if not(edges_array.shape[1] == 2 or edges_array.shape[1] == 3):
+			raise ValueError("edges must be a list of lists or tuples of the same length (2 or 3), where the third column indicates the color")
+
+		self.graph = _nx.MultiGraph()
+		self.graph.add_edges_from(edges_array)
+
+		super().__init__()
+		
+	def get_edges_from_color(self, color):
+		# Could be handy to get all edges from a color
+		return NotImplementedError
+
+	@property
+	def adjacency_list(self):
+		return [list(self.graph.neighbors(node)) for node in self.graph.nodes]
+
+	@property
+	def is_connected(self):
+		# TODO: how to check if a multigraph is connected?
+		return _nx.is_connected(self.graph)
+
+	@property
+	def edges(self):
+		return [e for e in self.graph.edges()]
```

Comment:
can't you just return self.graph.edges ?

Reply:
Yes, but as this is a MultiGraph, the list that is returned only contains pairs. self.graph.edges contains triplets (with the respective colour for each edge).

Reply:
The `networkx` interface is... interesting here:
```python
>>> g = nx.MultiGraph()
>>> g.add_edges_from([(0, 1, 0), (1, 2, 1), (2, 3, 0), (3, 0, 1)])
>>> [e for e in g.edges]
[(0, 1, 0), (0, 3, 1), (1, 2, 1), (2, 3, 0)]
>>> [e for e in g.edges()]
[(0, 1), (0, 3), (1, 2), (2, 3)]
```
Iterating over `.edges` gives triplets, iterating over `.edges()` gives pairs (unless called as `.edges(keys=True)`, the it also includes the colors/keys).

Maybe we should actually just include the colors as well when returning the edges. Having a separate `edge_colors` property (which actually does not return just the colors, so the name is not perfect) might be unnecessary. What do you think?